src/asr/asr_pytorch.py

```diff
@@ -411,55 +411,81 @@ def remove_dataparallel(state_dict):
-        y_true = map(int, recog_json[name]['output'][0]['tokenid'].split())
-
-        # print out decoding result
-        seq_hat = [train_args.char_list[int(idx)] for idx in y_hat]
-        seq_true = [train_args.char_list[int(idx)] for idx in y_true]
-        seq_hat_text = "".join(seq_hat).replace('<space>', ' ')
-        seq_true_text = "".join(seq_true).replace('<space>', ' ')
-        logging.info("groundtruth[%s]: " + seq_true_text, name)
-        logging.info("prediction [%s]: " + seq_hat_text, name)
-
-        # copy old json info
-        new_json[name] = dict()
-        new_json[name]['utt2spk'] = recog_json[name]['utt2spk']
-
-        # added recognition results to json
-        logging.debug("dump token id")
-        out_dic = dict()
-        for _key in recog_json[name]['output'][0]:
-            out_dic[_key] = recog_json[name]['output'][0][_key]
-
-        # TODO(karita) make consistent to chainer as idx[0] not idx
-        out_dic['rec_tokenid'] = " ".join([str(idx) for idx in y_hat])
-        logging.debug("dump token")
-        out_dic['rec_token'] = " ".join(seq_hat)
-        logging.debug("dump text")
-        out_dic['rec_text'] = seq_hat_text
-
-        new_json[name]['output'] = [out_dic]
-        # TODO(nelson): Modify this part when saving more than 1 hyp is enabled
-        # add n-best recognition results with scores
-        if args.beam_size > 1 and len(nbest_hyps) > 1:
-            for i, hyp in enumerate(nbest_hyps):
-                y_hat = hyp['yseq'][1:]
-                seq_hat = [train_args.char_list[int(idx)] for idx in y_hat]
-                seq_hat_text = "".join(seq_hat).replace('<space>', ' ')
-                new_json[name]['rec_tokenid' + '[' + '{:05d}'.format(i) + ']'] = " ".join([str(idx) for idx in y_hat])
-                new_json[name]['rec_token' + '[' + '{:05d}'.format(i) + ']'] = " ".join(seq_hat)
-                new_json[name]['rec_text' + '[' + '{:05d}'.format(i) + ']'] = seq_hat_text
-                new_json[name]['score' + '[' + '{:05d}'.format(i) + ']'] = hyp['score']
+    batchsize = args.batchsize
+
+    try:
+        from itertools import zip_longest as zip_longest
+    except:
+        from itertools import izip_longest as zip_longest
+    def grouper(n, iterable, fillvalue=None):
+        args = [iter(iterable)] * n
+        return zip_longest(*args, fillvalue=fillvalue)
+
+    for names in grouper(batchsize, recog_json.keys(), None):
+        names = [name for name in names if name]
+        feats = [kaldi_io_py.read_mat(recog_json[name]['input'][0]['feat'])
+                 for name in names]
+
+        sorted_index = sorted(range(len(feats)), key=lambda i: -len(feats[i]))
+        names = [names[i] for i in sorted_index]
+        feats = [feats[i] for i in sorted_index]
+        y_true = [map(int, recog_json[name]['output'][0]['tokenid'].split())
+                  for name in names]
+
+        nbest_hyps = e2e.recognize_batch(feats, args, train_args.char_list, rnnlm=rnnlm)
```

Comment:
You should replace `e2e.recognize_batch` -> `e2e.recognize`. It is not easy to maintain two recognition code. Also make sure that the result becomes equivalent to the original when the beamsize = 1